Add tablet merge support

raphaelsc opened the pull request:
The goal of merge is to reduce the tablet count for a shrinking table. Similar to how split increases the count while the table is growing. The load balancer decision to merge is implemented today (came with infrastructure introduced for split), but it wasn't handled until now.

Initial tablet count is respected while the table is in "growing mode". For example, the table leaves it if there was a need to split above the initial tablet count. After the table leaves the mode, the average size can be trusted to determine that the table is shrinking. Merge decision is emitted if the average tablet size is 50% of the target. Hysteresis is applied to avoid oscillations between split and merges.

Similar to split, the decision to merge is recorded in tablet map's resize_type field with the string "merge". This is important in case of coordinator failover, so new coordinator continues from where the old left off.

Unlike split, the preparation phase during merge is not done by the replica (with split compactions), but rather by the coordinator by co-locating sibling tablets in the same node's shard. We can define sibling tablets as tablets that have contiguous range and will become one after merge. The concept is based on the power-of-two constraint and token contiguity. For example, in a table with 4 tablets, tablets of ids 0 and 1 are siblings, 2 and 3 are also siblings.

The algorithm for co-locating sibling tablets is very simple. The balancer is responsible for it, and it will emit migrations so that "odd" tablet will follow the "even" one. For example, tablet 1 will be migrated to where tablet 0 lives. Co-location is low in priority, it's not the end of the world to delay merge, but it's not ideal to delay e.g. decommission or even regular load balancing as that can translate into temporary unbalancing, impacting the user activities. So co-location migrations will happen when there is no more important work to do.
While regular balancing is higher in priority, it will not undo the co-location work done so far. It does that by treating co-located tablets as if they were already merged. The load inversion convergence check was adjusted so balancer understand when two tablets are being migrated instead of one, to avoid oscillations.

When balancer completes co-location work for a table undergoing merge, it will put the id of the table into the resize_plan, which is about communicating with the topology coordinator that a table is ready for it. With all sibling tablets co-located, the coordinator can resize the tablet map (reduce it by a factor of 2) and record the new map into group0. All the replicas will react to it (on token metadata update) by merging the storage (memtable(s) + sstables) of sibling tablets into one.

Fixes #18181.

system test details:

test: https://github.com/pehala/scylla-cluster-tests/blob/tablets_split_merge/tablets_split_merge_test.py
yaml file: https://github.com/pehala/scylla-cluster-tests/blob/tablets_split_merge/test-cases/features/tablets/tablets-split-merge-test.yaml

instance type: i3.8xlarge
nodes: 3
target tablet size: 0.5G (scaled down by 10, to make it easier to trigger splits and merges)
description: multiple cycles of growing and shrinking the data set in order to trigger splits and merges.
data_set_size: ~100G
initial_tablets: 64, so it grew to 128 tablets on split, and back to 64 on merge.

latency of reads and writes that happened in parallel to split and merge:
```
$ for i in scylla-bench*; do cat $i | grep "Mode\|99th:\|99\.9th:"; done
Mode:			 write
  99.9th:	 3.145727ms 
  99th:		 1.998847ms 
  99.9th:	 3.145727ms 
  99th:		 2.031615ms 
Mode:			 read
  99.9th:	 3.145727ms 
  99th:		 2.031615ms 
  99.9th:	 3.145727ms 
  99th:		 2.031615ms 
Mode:			 write
  99.9th:	 3.047423ms 
  99th:		 1.933311ms 
  99.9th:	 3.047423ms 
  99th:		 1.933311ms 
Mode:			 read
  99.9th:	 3.145727ms 
  99th:		 1.900543ms 
  99.9th:	 3.145727ms 
  99th:		 1.900543ms 
Mode:			 write
  99.9th:	 5.079039ms 
  99th:		 3.604479ms 
  99.9th:	 35.389439ms 
  99th:		 25.624575ms 
Mode:			 write
  99.9th:	 3.047423ms 
  99th:		 1.998847ms 
  99.9th:	 3.047423ms 
  99th:		 1.998847ms 
Mode:			 read
  99.9th:	 3.080191ms 
  99th:		 2.031615ms 
  99.9th:	 3.112959ms 
  99th:		 2.031615ms 
```

raphaelsc commented:
@avikivity @tgrabiec ping for review.

avikivity commented on locator/tablets.hh:
```diff
@@ -61,7 +61,11 @@ struct tablet_replica {
+
+struct tablet_replica_less_comparator {
+    bool operator()(const tablet_replica& r1, tablet_replica& r2) const noexcept { return r1 < r2; }
 };
```
Not really needed, std::less<> will work, or std::less\<tablet_replica>.

avikivity commented on locator/tablets.hh:
```diff
@@ -278,6 +278,8 @@ struct resize_decision {
     sstring type_name() const;
     seq_number_t next_sequence_number() const;
+    // Returns true if this is the initial decision, before split or merge was emitted.
+    bool initial_decision() const;
```
Why is this important?

raphaelsc replied:
I will add motivation to changelog as you requested later, in v2.

avikivity commented on locator/tablets.hh:
```diff
@@ -403,6 +409,10 @@ public:
     future<> for_each_tablet(seastar::noncopyable_function<future<>(tablet_id, const tablet_info&)> func) const;
 
+    /// Calls a given function for each sibling tablet in the map in token ownership order.
+    /// If tablet count == 1, then there will be only one call and 2nd tablet_desc is disengaged.
```
Maybe there should be zero calls? After all there are no pairs.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -147,6 +147,51 @@ struct migration_badness {
+template <class... Ts>
+struct overloaded : Ts... {
+    using Ts::operator()...;
+};
```
We have utils::overloaded_functor.

raphaelsc replied:
Nice to know.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -147,6 +147,51 @@ struct migration_badness {
+    using Ts::operator()...;
+};
+
+struct tablet_candidate {
```
tablet_merge_candidate?

raphaelsc replied:
The reason I preferred to name it tablet_candidate (I also thought of balancer_candidate) is because we're using this not only for co-location work, but also for regular balancing. Once wired, it will be what the balancer will be working with, even when the context is not merge.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -614,6 +614,12 @@ class load_balancer {
 
+    tablet_replica_set sorted_replicas_for_tablet_load(const tablet_info& ti, const tablet_transition_info* trinfo) const {
+        auto set = get_replicas_for_tablet_load(ti, trinfo);
+        std::ranges::sort(set, tablet_replica_less_comparator());
```
The comparator parameter is unnecessary, the default std::less will work.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -497,8 +497,16 @@ class load_balancer {
+            // a table is shrinking and requires merge.
+            // FIXME: this is not perfect and we may want to leave the mode too if we detect
+            //  average size is decreasing significantly, before any split happened.
+            bool left_growing_mode = !d.resize_decision.initial_decision();
```
Please document the reason for initial_decision in its patch changelog.

raphaelsc replied:
Sure.

avikivity commented on service/topology_state_machine.cc:
```diff
@@ -147,7 +147,7 @@ static std::unordered_map<topology::transition_state, sstring> transition_state_
     {topology::transition_state::write_both_read_new, "write both read new"},
     {topology::transition_state::tablet_migration, "tablet migration"},
-    {topology::transition_state::tablet_split_finalization, "tablet split finalization"},
+    {topology::transition_state::tablet_resize_finalization, "tablet resize finalization"},
```
Is this a compatibility problem?

raphaelsc replied:
good catch, we do reverse lookup using string to transition id, when loading topology state. I can add tablet_merge_finalization and do this:
```
            case topology::transition_state::tablet_split_finalization:
            case topology::transition_state::tablet_merge_finalization:
                co_await handle_tablet_resize_finalization(std::move(guard));
```

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -715,6 +717,99 @@ class load_balancer {
+                if (!t2_opt) {
+                    on_internal_error(lblogger, format("Unable to find sibling tablet during co-location, with tablet count {}, for table {}",
+                                                       tmap.tablet_count(), table));
+                }
```
May as well not iterate at all for single tablets.

raphaelsc replied:
This works as an assertion, since balancer should never emit a merge decision for a single-tablet table. Also that tablet count must be power-of-two. It's an invariant today, but I thought it doesn't hurt to check, works as a documentation too. In practice, we shouldn't bump into this unless we have a terrible bug.

avikivity commented:
Looks good, very readable series.

I'm worried that the long vectors/sorts will be problematic on large tables. But this is not new with the series, we'll need a pass to soften it for large tables.

Please enrich the synthetic tests with merge scenarios, for example test/perf/tablet_load_balancing.cc.

I'd also like to see a scylla-cluster-test run, though I don't think we'll have scenarios that trigger it. Say a table that grows to 1 TB, shrinks to 200GB, grows again to 1 TB, shrinks to 0 TB. Perhaps via carefully planted TTLs.

raphaelsc commented:
Changes in v2:

- Replaced tablet_replica_less_comparator by std::less<tablet_replica>
- Added motivation explaining 'locator/tablets: Introduce resize_decision::initial_decision()'
- Use functor in utils/overloaded_functor.hh instead of adding new one
- In patch that renames transition state, lookup of the state using deprecated name "tablet split finalization" was fixed.
- Complete handling of tablet merge in replica layer (merge additional groups into main one in background).
- Added new test to verify correctness of balancer decisions (including colocation and regular migrations) during merge, using random scenarios
- With the new test, a few bugs in co-location were found and fixed (see 'service: Co-locate sibling tablets for a table undergoing merge').
- Wrote missing doc in docs/dev/topology-over-raft.md

NOTE: I am currently working with Yaron to stress merge using longevity.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12434/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12434/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_experimental_raft/test_tablets
#### Failed Tests (1/2733):
- [test_tablet_merge.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12434/testReport/junit/%28root%29/test_tablets/test_tablet_merge_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge.debug.9)
#### Build Details:
- Duration: 2 hr 4 min
- Builder: spider2.cloudius-systems.com

raphaelsc commented:
Changes in v3:
- Update compaction group id on merge completion, since old id is invalid and cannot be used to look up tablet map

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12475/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12475/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12475/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12275/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1184/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4251/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12475/)
#### Build Details:
- Duration: 8 hr 28 min
- Builder: spider1.cloudius-systems.com

tgrabiec commented:
Reviewing.

raphaelsc commented:
> Reviewing.

@tgrabiec ping.

avikivity commented:
Please decorate the cover letter with results of testing at scale.

raphaelsc commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -923,6 +927,114 @@ async def check():
+
+    await s1_log.wait_for("Emitting resize decision of type merge", from_mark=s1_mark)
+    # Waits for balancer to co-locate sibling tablets
+    time.sleep(5)
```
Yes, we can wait for co-location to be completely done here, rather than sleeping.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -711,6 +714,136 @@ class load_balancer {
+                auto ret = first_non_matching_replicas(r1, r2);
+                if (!ret) {
+                    return make_ready_future<>();
+                }
```
Yes, since we only got here if sibling tablet replicas aren't colocated.

tgrabiec replied:
Then shouldn't it be on_internal_error() ?

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -711,6 +714,136 @@ class load_balancer {
+                if (tmap.has_transitions()) {
+                    return make_ready_future<>();
+                }
+                if (migrating(t1) || migrating(t2)) {
```
As you noted later, the former check must be gone.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -816,14 +842,21 @@ class load_balancer {
 
+    // Precondition: all migration streaming info have same source and destination.
+    bool can_accept_load(node_load_map& nodes, const migration_streaming_info_vector& infos) {
+        // Since all migration info have source and destination, the load check can be easily done
```
I thought about that, but we'd need to copy node_load_map, since we'd have to apply the load incrementally on each iteration, so that the next iteration will account for the previous one. I am not completely happy with this either, but I thought it was fine to have this precondition and improve it later.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -383,7 +383,7 @@ class load_balancer {
     struct skipped_candidate {
         tablet_replica replica;
-        global_tablet_id tablet;
+        tablet_candidate tablet;
```
I remember I was willing to rename things to 'tablet_candidate candidate', but to reduce changes and make review easier I decided to keep it. If we rename it to tablet_candidate_set, then we name it either tablets or simply candidate. what about migration_candidate? even thought it might span two tablet replicas, we can think of it as a abstraction to some logical unit to be migrated. then existing migration_candidate needs to be renamed to migration_candidate_descriptor or simply migration_descriptor. what do you think?

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -2060,14 +2091,62 @@ class load_balancer {
+            };
 
-                for (auto&& replica : get_replicas_for_tablet_load(ti, trinfo)) {
+            // If a table is undergoing merge, co-located sibling tablets will be treated as a single migration candidate, even though
```
You're right, I had that in mind but my attempt to make it succint was not a good idea, since it is imprecise.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -1796,24 +1823,29 @@ class load_balancer {
+                                                        skip->second.tablet, src, nodes_to_drain, nodes_by_load_dst, replicas));
                     }
-                    src_node_info.skipped_candidates.emplace_back(src, source_tablet, std::move(skip->viable_targets));
+                    src_node_info.skipped_candidates.emplace_back(src, source_tablet, std::move(skip->first.viable_targets));
```
Ok. thanks.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -2060,14 +2091,62 @@ class load_balancer {
+                    return tid == t1.tid ? t1 : t2;
+                };
+
+                while (auto next = next_replica()) {
```
Ok.

raphaelsc commented on replica/table.cc:
```diff
@@ -2428,6 +2458,51 @@ future<> tablet_storage_group_manager::handle_tablet_split_completion(const loca
+                cg->update_id(new_tid);
+                new_sg->add_merging_group(cg);
+            });
+            // FIXME: we MUST schedule a background action to flush memtable and move sstables of merging groups into the main one.
```
I thought about this, but it is complicated since memtable_list has a lambda that references its underlying compaction_group, and ongoing flush takes compaction group as a reference (see table::make_memtable_list()). So if we just move the list, we end up with a dangling pointer. We cannot afford to wait for flush here either, since the changes in replica state must be applied atomically. This why I need merging groups. The groups will be eventually merged into main group of new tablet. Maybe we can optimize it later with some work in table::seal_active_memtable() where it works with a reference to memtable list, and grabs the compaction group only when needed (e.g. when adding the sstable produced to the set).

raphaelsc commented on replica/table.cc:
```diff
@@ -2458,6 +2517,34 @@ future<> tablet_storage_group_manager::handle_tablet_split_completion(const loca
+                auto main_group = sg.main_compaction_group();
+                for (auto& group : sg.merging_groups()) {
+                    // Flushes memtable, so all the data can be moved.
+                    co_await group->flush();
```
You're right. we must serialize with a pending write that sleeps in dirty_memory_region_group().run_when_memory_available(), inside table::do_apply(). Newer writes aren't problem since they pick a different group (not merging ones).

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -147,6 +148,46 @@ struct migration_badness {
 
+struct colocated_tablet {
+    global_tablet_id left_tablet;
+    global_tablet_id right_tablet;
```
But it's not important to optimize this, so disregard.

avikivity commented on test/boost/tablets_test.cc:
```diff
@@ -2600,6 +2618,112 @@ SEASTAR_THREAD_TEST_CASE(basic_tablet_storage_splitting_test) {
+
+    const uint64_t target_tablet_size = service::default_target_tablet_size;
+    auto size_to_trigger_merge = [&] () -> uint64_t {
+        return (target_tablet_size * 0.4f) * tablet_count();
```
Why is 0.4 and not 0.5?

raphaelsc replied:
> Why is 0.4 and not 0.5?

I think my intention was to have size_that_certainly_triggers_merge, but I think it's indeed better to have maximum_size_that_triggers_merge.

avikivity commented:
How long does the test take to run?

raphaelsc commented:
> How long does the test take to run?

30s in debug mode, but longer tests in topology_experimental_raft.test_tablets take more than 45s. I think I can remove one 5s sleep and reduce it more. 

src: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12475/artifact/testlog/x86_64/debug/xml/topology_experimental_raft.test_tablets.1.xunit.xml

raphaelsc commented:
Changes in v4:
- rebased
- dropped extra spaces in tablet-mon
- take into account that constraints (viable targets) of co-located replicas might be different.
    see patch "service: Add replication constraint check variant for ..."
- rename existing migration_candidate (to a more descriptive name), and use migration_candidate for representing either a single replica or co-located ones.
- fix a possible race of merge completion fiber with blocked writes to compaction group that will be merged and discarded
- restored replica order for not messing with base-view pairing. still cannot handle RF=N, will be a follow-up with idea outlined in https://github.com/scylladb/scylladb/issues/17265

raphaelsc commented on test/boost/tablets_test.cc:
```diff
@@ -1347,11 +1347,23 @@ void apply_resize_plan(token_metadata& tm, const migration_plan& plan) {
+        auto new_tmap = co_await talloc.resize_tablets(tm, table_id);
+        auto new_resize_decision = locator::resize_decision{};
+        new_resize_decision.sequence_number = old_tmap.resize_decision().next_sequence_number();
+        new_tmap.set_resize_decision(std::move(new_resize_decision));
```
pre-existing, so postponing.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13198/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13198/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/251/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4801/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1727/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12823/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13198/)
#### Failed Tests (1/45689):
- [test_random_failures[stop_after_setting_mode_to_normal_raft_topology-restart_non_coordinator_node].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13198/testReport/junit/%28root%29/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_setting_mode_to_normal_raft_topology_restart_non_coordinator_node__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_setting_mode_to_normal_raft_topology-restart_non_coordinator_node].debug.2)
#### Build Details:
- Duration: 10 hr
- Builder: spider7.cloudius-systems.com

avikivity commented:
> Please decorate the cover letter with results of testing at scale.

Still missing

raphaelsc commented:
> > Please decorate the cover letter with results of testing at scale.
> 
> Still missing

we're working on it https://github.com/scylladb/scylla-cluster-tests/pull/8948.

raphaelsc commented:
> > Please decorate the cover letter with results of testing at scale.
> 
> Still missing

@avikivity @tgrabiec Botond suggested that we initially disallow merge for tables containing views. It's a good idea I think, since it's better to have something working optimally for one thing, than somethig working partially for everything. To preserve order while not implementing the complex idea by Tomek, it means co-location efficiency depends on topology (for example today it's optimal when RF=#_of_racks, but it may fail under single rack since Tomek and I agreed on postponing its handling due to complexity).

avikivity commented:
> > > Please decorate the cover letter with results of testing at scale.
> > 
> > 
> > Still missing
> 
> @avikivity @tgrabiec Botond suggested that we initially disallow merge for tables containing views. It's a good idea I think, since it's better to have something working optimally for one thing, than somethig working partially for everything. To preserve order while not implementing the complex idea by Tomek, it means co-location efficiency depends on topology (for example today it's optimal when RF=#_of_racks, but it may fail under single rack since Tomek and I agreed on postponing its handling due to complexity).

Sure

raphaelsc commented:
Changes in v5:
- Rebased
- Restrict merge to tables without views, since preserving replica order results in co-location uncertainty under certain topologies. Limitation will later be addressed with https://github.com/scylladb/scylladb/issues/17265.
- Added tests: 'test_load_balancing_merge_colocation_with_single_rack', 'test_load_balancing_merge_colocation_with_decomission'
- Moved test_tablet_merge to its own .py file, to favor concurrency

raphaelsc commented:
scale test result:

multiple cycles of splits and merges, to 128 tablets back to 64.

![image](https://github.com/user-attachments/assets/f603b4a9-b36c-444d-bd1e-8bc7c661e1e1)
![image](https://github.com/user-attachments/assets/fc77ad14-0fd7-41aa-a435-32fa82cd5bf3)


```
$ for i in scylla-bench*; do cat $i | grep "Mode\|99th:\|99\.9th:"; done
Mode:			 write
  99.9th:	 3.145727ms 
  99th:		 1.998847ms 
  99.9th:	 3.145727ms 
  99th:		 2.031615ms 
Mode:			 read
  99.9th:	 3.145727ms 
  99th:		 2.031615ms 
  99.9th:	 3.145727ms 
  99th:		 2.031615ms 
Mode:			 write
  99.9th:	 3.047423ms 
  99th:		 1.933311ms 
  99.9th:	 3.047423ms 
  99th:		 1.933311ms 
Mode:			 read
  99.9th:	 3.145727ms 
  99th:		 1.900543ms 
  99.9th:	 3.145727ms 
  99th:		 1.900543ms 
Mode:			 write
  99.9th:	 5.079039ms 
  99th:		 3.604479ms 
  99.9th:	 35.389439ms 
  99th:		 25.624575ms 
Mode:			 write
  99.9th:	 3.047423ms 
  99th:		 1.998847ms 
  99.9th:	 3.047423ms 
  99th:		 1.998847ms 
Mode:			 read
  99.9th:	 3.080191ms 
  99th:		 2.031615ms 
  99.9th:	 3.112959ms 
  99th:		 2.031615ms 
```

/cc @avikivity

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13329/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13329/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_merge
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/339/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12950/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4937/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1840/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13329/)
#### Build Details:
- Duration: 10 hr
- Builder: spider4.cloudius-systems.com

avikivity commented:
> scale test result:
> 
> multiple cycles of splits and merges, to 128 tablets back to 64.
> 
> ![image](https://private-user-images.githubusercontent.com/1409139/389128739-f603b4a9-b36c-444d-bd1e-8bc7c661e1e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgzODksIm5iZiI6MTczMjQ1ODA4OSwicGF0aCI6Ii8xNDA5MTM5LzM4OTEyODczOS1mNjAzYjRhOS1iMzZjLTQ0NGQtYmQxZS04YmM3YzY2MWUxZTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjRUMTQyMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ0MjFhYWQzNDg0OTFlMmVlYjc0ZWFlMjliMTYwMGJlNjIxNzAzMDMzZjRjN2E4NjljYTY2YWU4YjNmMzQ2ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.-WSyoq88W3KqtrsFgsw-wOrT23Udjpb6ibgVuRKKj68) ![image](https://private-user-images.githubusercontent.com/1409139/389132266-fc77ad14-0fd7-41aa-a435-32fa82cd5bf3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgzODksIm5iZiI6MTczMjQ1ODA4OSwicGF0aCI6Ii8xNDA5MTM5LzM4OTEzMjI2Ni1mYzc3YWQxNC0wZmQ3LTQxYWEtYTQzNS0zMmZhODJjZDViZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjRUMTQyMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTVhYjZmZGQxNjIzOTczMTczMjZmODA0ODBjYWNmNTU5NWJjOTUxNGJhM2I0OTc2MGJhNmIyZTE5OTc4NDQ2NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.KmD697bmC60okrlCDzIHEmy6obPQOTTcZBC_2sVDxOE)
> 
> ```
> $ for i in scylla-bench*; do cat $i | grep "Mode\|99th:\|99\.9th:"; done
> Mode:			 write
>   99.9th:	 3.145727ms 
>   99th:		 1.998847ms 
>   99.9th:	 3.145727ms 
>   99th:		 2.031615ms 
> Mode:			 read
>   99.9th:	 3.145727ms 
>   99th:		 2.031615ms 
>   99.9th:	 3.145727ms 
>   99th:		 2.031615ms 
> Mode:			 write
>   99.9th:	 3.047423ms 
>   99th:		 1.933311ms 
>   99.9th:	 3.047423ms 
>   99th:		 1.933311ms 
> Mode:			 read
>   99.9th:	 3.145727ms 
>   99th:		 1.900543ms 
>   99.9th:	 3.145727ms 
>   99th:		 1.900543ms 
> Mode:			 write
>   99.9th:	 5.079039ms 
>   99th:		 3.604479ms 
>   99.9th:	 35.389439ms 
>   99th:		 25.624575ms 
> Mode:			 write
>   99.9th:	 3.047423ms 
>   99th:		 1.998847ms 
>   99.9th:	 3.047423ms 
>   99th:		 1.998847ms 
> Mode:			 read
>   99.9th:	 3.080191ms 
>   99th:		 2.031615ms 
>   99.9th:	 3.112959ms 
>   99th:		 2.031615ms 
> ```
> 
> /cc @avikivity

The test report should be in the cover letter, and describe what workload was running, not just latency at nanosecond resolution.

I think we should have scale-in/scale-out in parallel with split/merge, to make sure they work together.

raphaelsc commented:
> > scale test result:
> > 
> > multiple cycles of splits and merges, to 128 tablets back to 64.
> > 
> > ![image](https://private-user-images.githubusercontent.com/1409139/389128739-f603b4a9-b36c-444d-bd1e-8bc7c661e1e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgzODksIm5iZiI6MTczMjQ1ODA4OSwicGF0aCI6Ii8xNDA5MTM5LzM4OTEyODczOS1mNjAzYjRhOS1iMzZjLTQ0NGQtYmQxZS04YmM3YzY2MWUxZTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjRUMTQyMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ0MjFhYWQzNDg0OTFlMmVlYjc0ZWFlMjliMTYwMGJlNjIxNzAzMDMzZjRjN2E4NjljYTY2YWU4YjNmMzQ2ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.-WSyoq88W3KqtrsFgsw-wOrT23Udjpb6ibgVuRKKj68) ![image](https://private-user-images.githubusercontent.com/1409139/389132266-fc77ad14-0fd7-41aa-a435-32fa82cd5bf3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgzODksIm5iZiI6MTczMjQ1ODA4OSwicGF0aCI6Ii8xNDA5MTM5LzM4OTEzMjI2Ni1mYzc3YWQxNC0wZmQ3LTQxYWEtYTQzNS0zMmZhODJjZDViZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjRUMTQyMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTVhYjZmZGQxNjIzOTczMTczMjZmODA0ODBjYWNmNTU5NWJjOTUxNGJhM2I0OTc2MGJhNmIyZTE5OTc4NDQ2NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.KmD697bmC60okrlCDzIHEmy6obPQOTTcZBC_2sVDxOE)
> > 
> > ```
> > $ for i in scylla-bench*; do cat $i | grep "Mode\|99th:\|99\.9th:"; done
> > Mode:			 write
> >   99.9th:	 3.145727ms 
> >   99th:		 1.998847ms 
> >   99.9th:	 3.145727ms 
> >   99th:		 2.031615ms 
> > Mode:			 read
> >   99.9th:	 3.145727ms 
> >   99th:		 2.031615ms 
> >   99.9th:	 3.145727ms 
> >   99th:		 2.031615ms 
> > Mode:			 write
> >   99.9th:	 3.047423ms 
> >   99th:		 1.933311ms 
> >   99.9th:	 3.047423ms 
> >   99th:		 1.933311ms 
> > Mode:			 read
> >   99.9th:	 3.145727ms 
> >   99th:		 1.900543ms 
> >   99.9th:	 3.145727ms 
> >   99th:		 1.900543ms 
> > Mode:			 write
> >   99.9th:	 5.079039ms 
> >   99th:		 3.604479ms 
> >   99.9th:	 35.389439ms 
> >   99th:		 25.624575ms 
> > Mode:			 write
> >   99.9th:	 3.047423ms 
> >   99th:		 1.998847ms 
> >   99.9th:	 3.047423ms 
> >   99th:		 1.998847ms 
> > Mode:			 read
> >   99.9th:	 3.080191ms 
> >   99th:		 2.031615ms 
> >   99.9th:	 3.112959ms 
> >   99th:		 2.031615ms 
> > ```
> > 
> > /cc @avikivity
> 
> The test report should be in the cover letter, and describe what workload was running, not just latency at nanosecond resolution.

The report was already in the lette. I thought the first paragraph made it clear that it had multiple cycles of growing and shrinking data set, causing the number of tablets to grow to 128 tablets and back to 64. Reads and writes in parallel. I will improve the description.

> 
> I think we should have scale-in/scale-out in parallel with split/merge, to make sure they work together.

I think we can add a py test that adds / remove nodes during merge too (a quicker way (not a replacement though) of testing merge correctness during topology ops, we already have such tests in place for split). Improvements to large scale test can come later and issues can be fixed down the road. If I add the py test we can merge series (it rots quickly) and continue testing. Are you fine with that?

raphaelsc commented:
> > scale test result:
> > 
> > multiple cycles of splits and merges, to 128 tablets back to 64.
> > 
> > ![image](https://private-user-images.githubusercontent.com/1409139/389128739-f603b4a9-b36c-444d-bd1e-8bc7c661e1e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgzODksIm5iZiI6MTczMjQ1ODA4OSwicGF0aCI6Ii8xNDA5MTM5LzM4OTEyODczOS1mNjAzYjRhOS1iMzZjLTQ0NGQtYmQxZS04YmM3YzY2MWUxZTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjRUMTQyMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ0MjFhYWQzNDg0OTFlMmVlYjc0ZWFlMjliMTYwMGJlNjIxNzAzMDMzZjRjN2E4NjljYTY2YWU4YjNmMzQ2ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.-WSyoq88W3KqtrsFgsw-wOrT23Udjpb6ibgVuRKKj68) ![image](https://private-user-images.githubusercontent.com/1409139/389132266-fc77ad14-0fd7-41aa-a435-32fa82cd5bf3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgzODksIm5iZiI6MTczMjQ1ODA4OSwicGF0aCI6Ii8xNDA5MTM5LzM4OTEzMjI2Ni1mYzc3YWQxNC0wZmQ3LTQxYWEtYTQzNS0zMmZhODJjZDViZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjRUMTQyMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTVhYjZmZGQxNjIzOTczMTczMjZmODA0ODBjYWNmNTU5NWJjOTUxNGJhM2I0OTc2MGJhNmIyZTE5OTc4NDQ2NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.KmD697bmC60okrlCDzIHEmy6obPQOTTcZBC_2sVDxOE)
> > 
> > ```
> > $ for i in scylla-bench*; do cat $i | grep "Mode\|99th:\|99\.9th:"; done
> > Mode:			 write
> >   99.9th:	 3.145727ms 
> >   99th:		 1.998847ms 
> >   99.9th:	 3.145727ms 
> >   99th:		 2.031615ms 
> > Mode:			 read
> >   99.9th:	 3.145727ms 
> >   99th:		 2.031615ms 
> >   99.9th:	 3.145727ms 
> >   99th:		 2.031615ms 
> > Mode:			 write
> >   99.9th:	 3.047423ms 
> >   99th:		 1.933311ms 
> >   99.9th:	 3.047423ms 
> >   99th:		 1.933311ms 
> > Mode:			 read
> >   99.9th:	 3.145727ms 
> >   99th:		 1.900543ms 
> >   99.9th:	 3.145727ms 
> >   99th:		 1.900543ms 
> > Mode:			 write
> >   99.9th:	 5.079039ms 
> >   99th:		 3.604479ms 
> >   99.9th:	 35.389439ms 
> >   99th:		 25.624575ms 
> > Mode:			 write
> >   99.9th:	 3.047423ms 
> >   99th:		 1.998847ms 
> >   99.9th:	 3.047423ms 
> >   99th:		 1.998847ms 
> > Mode:			 read
> >   99.9th:	 3.080191ms 
> >   99th:		 2.031615ms 
> >   99.9th:	 3.112959ms 
> >   99th:		 2.031615ms 
> > ```
> > 
> > /cc @avikivity
> 
> The test report should be in the cover letter, and describe what workload was running, not just latency at nanosecond resolution.
> 
> I think we should have scale-in/scale-out in parallel with split/merge, to make sure they work together.

@pehala let's please synchronize tomorrow on making these changes to the test

avikivity commented:
> > > scale test result:
> > > multiple cycles of splits and merges, to 128 tablets back to 64.
> > > ![image](https://private-user-images.githubusercontent.com/1409139/389128739-f603b4a9-b36c-444d-bd1e-8bc7c661e1e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgzODksIm5iZiI6MTczMjQ1ODA4OSwicGF0aCI6Ii8xNDA5MTM5LzM4OTEyODczOS1mNjAzYjRhOS1iMzZjLTQ0NGQtYmQxZS04YmM3YzY2MWUxZTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjRUMTQyMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ0MjFhYWQzNDg0OTFlMmVlYjc0ZWFlMjliMTYwMGJlNjIxNzAzMDMzZjRjN2E4NjljYTY2YWU4YjNmMzQ2ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.-WSyoq88W3KqtrsFgsw-wOrT23Udjpb6ibgVuRKKj68) ![image](https://private-user-images.githubusercontent.com/1409139/389132266-fc77ad14-0fd7-41aa-a435-32fa82cd5bf3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgzODksIm5iZiI6MTczMjQ1ODA4OSwicGF0aCI6Ii8xNDA5MTM5LzM4OTEzMjI2Ni1mYzc3YWQxNC0wZmQ3LTQxYWEtYTQzNS0zMmZhODJjZDViZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjRUMTQyMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTVhYjZmZGQxNjIzOTczMTczMjZmODA0ODBjYWNmNTU5NWJjOTUxNGJhM2I0OTc2MGJhNmIyZTE5OTc4NDQ2NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.KmD697bmC60okrlCDzIHEmy6obPQOTTcZBC_2sVDxOE)
> > > ```
> > > $ for i in scylla-bench*; do cat $i | grep "Mode\|99th:\|99\.9th:"; done
> > > Mode:			 write
> > >   99.9th:	 3.145727ms 
> > >   99th:		 1.998847ms 
> > >   99.9th:	 3.145727ms 
> > >   99th:		 2.031615ms 
> > > Mode:			 read
> > >   99.9th:	 3.145727ms 
> > >   99th:		 2.031615ms 
> > >   99.9th:	 3.145727ms 
> > >   99th:		 2.031615ms 
> > > Mode:			 write
> > >   99.9th:	 3.047423ms 
> > >   99th:		 1.933311ms 
> > >   99.9th:	 3.047423ms 
> > >   99th:		 1.933311ms 
> > > Mode:			 read
> > >   99.9th:	 3.145727ms 
> > >   99th:		 1.900543ms 
> > >   99.9th:	 3.145727ms 
> > >   99th:		 1.900543ms 
> > > Mode:			 write
> > >   99.9th:	 5.079039ms 
> > >   99th:		 3.604479ms 
> > >   99.9th:	 35.389439ms 
> > >   99th:		 25.624575ms 
> > > Mode:			 write
> > >   99.9th:	 3.047423ms 
> > >   99th:		 1.998847ms 
> > >   99.9th:	 3.047423ms 
> > >   99th:		 1.998847ms 
> > > Mode:			 read
> > >   99.9th:	 3.080191ms 
> > >   99th:		 2.031615ms 
> > >   99.9th:	 3.112959ms 
> > >   99th:		 2.031615ms 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > /cc @avikivity
> > 
> > 
> > The test report should be in the cover letter, and describe what workload was running, not just latency at nanosecond resolution.
> 
> The report was already in the lette. I thought the first paragraph made it clear that it had multiple cycles of growing and shrinking data set, causing the number of tablets to grow to 128 tablets and back to 64. Reads and writes in parallel. I will improve the description.

I have no idea how many nodes were involved, their size, the dataset size, CPU loading of the main group, what the workload does. Just some latency numbers.

It's like saying the average speed was 20 km/hour. For a run? a drive? a flight? What distance? nobody knows.

Ideally there should be enough information for someone to independently run the same test.

> 
> > I think we should have scale-in/scale-out in parallel with split/merge, to make sure they work together.
> 
> I think we can add a py test that adds / remove nodes during merge too (a quicker way (not a replacement though) of testing merge correctness during topology ops, we already have such tests in place for split). Improvements to large scale test can come later and issues can be fixed down the road. If I add the py test we can merge series (it rots quickly) and continue testing. Are you fine with that?

If we can test data movement in parallel with merging in test.py, that would be good. But remember test.py has to run quickly. Can we trigger merging independently of actual storage utilization? Perhaps adjusting the tablet size goal while the test is running is enough to trigger split/merge without huge amounts of writes.

raphaelsc commented:
> > > > scale test result:
> > > > multiple cycles of splits and merges, to 128 tablets back to 64.
> > > > ![image](https://private-user-images.githubusercontent.com/1409139/389128739-f603b4a9-b36c-444d-bd1e-8bc7c661e1e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgzODksIm5iZiI6MTczMjQ1ODA4OSwicGF0aCI6Ii8xNDA5MTM5LzM4OTEyODczOS1mNjAzYjRhOS1iMzZjLTQ0NGQtYmQxZS04YmM3YzY2MWUxZTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjRUMTQyMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ0MjFhYWQzNDg0OTFlMmVlYjc0ZWFlMjliMTYwMGJlNjIxNzAzMDMzZjRjN2E4NjljYTY2YWU4YjNmMzQ2ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.-WSyoq88W3KqtrsFgsw-wOrT23Udjpb6ibgVuRKKj68) ![image](https://private-user-images.githubusercontent.com/1409139/389132266-fc77ad14-0fd7-41aa-a435-32fa82cd5bf3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgzODksIm5iZiI6MTczMjQ1ODA4OSwicGF0aCI6Ii8xNDA5MTM5LzM4OTEzMjI2Ni1mYzc3YWQxNC0wZmQ3LTQxYWEtYTQzNS0zMmZhODJjZDViZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjRUMTQyMTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTVhYjZmZGQxNjIzOTczMTczMjZmODA0ODBjYWNmNTU5NWJjOTUxNGJhM2I0OTc2MGJhNmIyZTE5OTc4NDQ2NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.KmD697bmC60okrlCDzIHEmy6obPQOTTcZBC_2sVDxOE)
> > > > ```
> > > > $ for i in scylla-bench*; do cat $i | grep "Mode\|99th:\|99\.9th:"; done
> > > > Mode:			 write
> > > >   99.9th:	 3.145727ms 
> > > >   99th:		 1.998847ms 
> > > >   99.9th:	 3.145727ms 
> > > >   99th:		 2.031615ms 
> > > > Mode:			 read
> > > >   99.9th:	 3.145727ms 
> > > >   99th:		 2.031615ms 
> > > >   99.9th:	 3.145727ms 
> > > >   99th:		 2.031615ms 
> > > > Mode:			 write
> > > >   99.9th:	 3.047423ms 
> > > >   99th:		 1.933311ms 
> > > >   99.9th:	 3.047423ms 
> > > >   99th:		 1.933311ms 
> > > > Mode:			 read
> > > >   99.9th:	 3.145727ms 
> > > >   99th:		 1.900543ms 
> > > >   99.9th:	 3.145727ms 
> > > >   99th:		 1.900543ms 
> > > > Mode:			 write
> > > >   99.9th:	 5.079039ms 
> > > >   99th:		 3.604479ms 
> > > >   99.9th:	 35.389439ms 
> > > >   99th:		 25.624575ms 
> > > > Mode:			 write
> > > >   99.9th:	 3.047423ms 
> > > >   99th:		 1.998847ms 
> > > >   99.9th:	 3.047423ms 
> > > >   99th:		 1.998847ms 
> > > > Mode:			 read
> > > >   99.9th:	 3.080191ms 
> > > >   99th:		 2.031615ms 
> > > >   99.9th:	 3.112959ms 
> > > >   99th:		 2.031615ms 
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > /cc @avikivity
> > > 
> > > 
> > > The test report should be in the cover letter, and describe what workload was running, not just latency at nanosecond resolution.
> > 
> > 
> > The report was already in the lette. I thought the first paragraph made it clear that it had multiple cycles of growing and shrinking data set, causing the number of tablets to grow to 128 tablets and back to 64. Reads and writes in parallel. I will improve the description.
> 
> I have no idea how many nodes were involved, their size, the dataset size, CPU loading of the main group, what the workload does. Just some latency numbers.
> 
> It's like saying the average speed was 20 km/hour. For a run? a drive? a flight? What distance? nobody knows.
> 
> Ideally there should be enough information for someone to independently run the same test.

You're right.

> 
> > > I think we should have scale-in/scale-out in parallel with split/merge, to make sure they work together.
> > 
> > 
> > I think we can add a py test that adds / remove nodes during merge too (a quicker way (not a replacement though) of testing merge correctness during topology ops, we already have such tests in place for split). Improvements to large scale test can come later and issues can be fixed down the road. If I add the py test we can merge series (it rots quickly) and continue testing. Are you fine with that?
> 
> If we can test data movement in parallel with merging in test.py, that would be good. But remember test.py has to run quickly. Can we trigger merging independently of actual storage utilization? Perhaps adjusting the tablet size goal while the test is running is enough to trigger split/merge without huge amounts of writes.

we already kind of do scale model testing by adjusting tablet size goal proportionally in existing tests. we can finish it under a minute, and run multiple cycles of splits and merge, with topology ops in parallel to them. it's a good test to have in test.py I think since it stresses that scenario in a relatively short amount of time.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -699,6 +700,9 @@ class load_balancer {
         const locator::topology& topo = _tm->get_topology();
         migration_plan plan;
 
+        // Prepare table-wide resize plans first, since DC plan might have something to merge into the resize plan.
```
What might DC plan merge into the resize plan? Why can't we merge resize plan into the DC plan? Merging should be commutative, and DC plan making doesn't have visibility into the resize plan.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -877,6 +879,158 @@ class load_balancer {
+                if (r1 == r2) {
+                    return make_ready_future<>();
+                }
+                all_colocated = false;
```
get_replicas() work only with dc-local subset, but finalization takes effect for all dcs, so we shouldn't base the check on dc-local colocation.

raphaelsc replied:
good catch.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -877,6 +879,158 @@ class load_balancer {
+                    if (src.host == dst.host && src.shard != dst.shard) {
+                        return false;
+                    }
+                    return r2_hosts.contains(dst.host);
```
This doesn't check for rack overload, which means we may temporarily overload racks until all is colocated. We can probably live with it for now. @bhalevy

raphaelsc replied:
We can possibly improve it by sorting replica set by rack id. So with RF=#_of_racks, co-location should not increase  load on any rack.

raphaelsc replied:
Postponing this since it requires some more careful thought.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -149,6 +150,51 @@ struct migration_badness {
+// candidate. When candidate represents co-located replicas, it means that
+// the balancer will work to preserve the co-location by migrating those replicas
+// to same destination.
+struct migration_candidate {
```
migration_tablet_set tablet_set could be good too. migration_tablet_set tablets makes it sound like there will always be multiple tablets in each candidate. 

See that each candidate might span one or more tablets, which is why I decided to name it migration_candidate, represents a logical candidate, since it might span two physical entities (tablets).

Otherwise, we'll have each node having candidates being typed as a set of migration_tablet_set (not very easy to understand why there's a set of a set in candidate list). And skipped candidate will also store a migration_tablet_set. skipped candidate becomes a logical term anyway, since it will store a tablet_set, which is in reality it can be two migration candidates (tablets).

tgrabiec replied:
> migration_tablet_set tablet_set could be good too. migration_tablet_set tablets makes it sound like there will always be multiple tablets in each candidate.

plural doesn't imply that there must be more than one, it means there may be many. e.g. `_tables`, `_tablets`.  We can have 1 table and 1 tablet.

raphaelsc replied:
> > migration_tablet_set tablet_set could be good too. migration_tablet_set tablets makes it sound like there will always be multiple tablets in each candidate.
> 
> plural doesn't imply that there must be more than one, it means there may be many. e.g. `_tables`, `_tablets`. We can have 1 table and 1 tablet.

This is a good point. It's very common in our field to do that naming when using sets.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -990,6 +1016,16 @@ class load_balancer {
+    // Precondition: all migration streaming info have same source and destination.
+    //  FIXME: remove precondition but it's not easy without copying noad_load_map.
+    bool can_accept_load(node_load_map& nodes, const migration_streaming_info_vector& infos) {
+        // Since all migration info have source and destination, the load check can be easily done
```
will change , thanks.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -2002,28 +2029,36 @@ class load_balancer {
                     continue;
                 }
             }
-
```
Ah removed it by accident, will bring it back.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -2532,11 +2532,48 @@ class tablet_allocator_impl : public tablet_allocator::impl
+                                                old_left_tid, left_tablet_replicas, old_right_tid, right_tablet_replicas));
+            }
+
+            new_tablets.set_tablet(tid, left_tablet_info);
```
Ah I forget to address this by not throwing away current tablet info, but rather preserving its fields when building new tablet info.

tgrabiec replied:
We cannot merge tasks in general, since if left and right might have different task ids and then there's no way to merge them - we can store only one task per tablet. So merge finalization should proceed only if all tasks are mergeable. We can start with something simple and not merge if there are any tasks.

raphaelsc replied:
Ah I remember when you raised this to me, we were going with a timestamp or something, so I'd take a minimum among infos, but apparently we changed that.

tgrabiec replied:
There is still `repair_time` which we should merge, yes. `repair_task_info` is another problem.

raphaelsc replied:
why we cannot merge repair tasks if tablets are co-located? it should be fine as long as repair works with token ranges, and doesn't require tablet id stability. I am proceeding with this constraint, but let me know it indeed exists.

raphaelsc replied:
I don't know exactly how repair tasks are implemented, but we shouldn't have any active repair task when state machine transition into resize finalization state, no?

raphaelsc replied:
anyway, I will only emit the request to finalize when all sibling tablets have their infos mergeable.

tgrabiec replied:
> why we cannot merge repair tasks if tablets are co-located? it should be fine as long as repair works with token ranges, and doesn't require tablet id stability. I am proceeding with this constraint, but let me know it indeed exists.

left tablet may have a different task id assigned than the right tablet, so they belong to different jobs. Merging would mean that one of the tasks would be dropped and no-oped, which is not correct. The dropped task may be considered as finished without doing any work.

tgrabiec replied:
> I don't know exactly how repair tasks are implemented, but we shouldn't have any active repair task when state machine transition into resize finalization state, no?

We shouldn't have active repair transitions, but we may still have inactive (to be scheduled) tasks assigned. They can be assigned at any point in time in group0 history, even after state machine entered split finalization.

tgrabiec replied:
> anyway, I will only emit the request to finalize when all sibling tablets have their infos mergeable.

OK

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -312,7 +312,7 @@ class load_balancer {
     struct skipped_candidate {
         tablet_replica replica;
-        global_tablet_id tablet;
+        global_tablet_id candidate;
```
> migration_candidate -> logical candidate
> migration_candidate_info -> info about a migration candidate.

The new migration_candidate is not a full migration candidate. We move tablet replicas, but the new migration_candidate selects the tablet, not a particular replica. It makes sense when migration_candidate sits in the per-shard candidate list, since then it identifies tablet replica. But not when it sits in migration_candidate_info. 

I don't like the "_info" part because it's meaningless (what info?) as a distinctor. The names migration_candidate and migration_candidate_info are too similar. Perhaps a better name would be just "migration". migration_candidate_info selects the source and destination, so it's a full migration descriptor. Then the new "migration_candidate" could be salvaged.

tgrabiec replied:
Though "migration_candidate" still wouldn't seem accurate, since it doesn't always select the replica. It depends on context. So "migration_tablet_set" seems still more accurate to me.

raphaelsc replied:
Sure, will proceed that way.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -1943,14 +2002,18 @@ class load_balancer {
+
+                auto skip = check_constraints(nodes, tmap, src_node_info, nodes[dst.host], candidate, src_node_info.drained);
+                if (skip) {
+                    process_skip_info(candidate, std::move(*skip));
```
actually we weren't using migration_candidate yet, but rather candidate which is still tablet_id, so preexisting check_constraints() variant is picked, not new one introduced in the patch

raphaelsc commented:
Changes in v6:
- rename migration_candidate to migration_tablet_set
- extract sibling tablets replica set processing logic for improving readability
- don't finalize merge when tablet_info of sibling tablets cannot be merged (due to active repair task)
- only finalize merge when all DCs agree on co-location of a table
- added test that verify correctness of split and merge during topology changes

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13506/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13506/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_merge
#### Failed Tests (2/1035):
- [test_tablet_merge_simple.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13506/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.5)
- [test_tablet_split_and_merge_with_concurrent_topology_changes.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13506/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_split_and_merge_with_concurrent_topology_changes_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_split_and_merge_with_concurrent_topology_changes.debug.8)
#### Build Details:
- Duration: 1 hr 18 min
- Builder: spider2.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13532/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13532/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_merge
#### Failed Tests (1/4140):
- [test_tablet_split_and_merge_with_concurrent_topology_changes.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13532/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_split_and_merge_with_concurrent_topology_changes_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_split_and_merge_with_concurrent_topology_changes.debug.1)
#### Build Details:
- Duration: 7 hr 34 min
- Builder: i-0e52977c1232f92bd (m5ad.12xlarge)

raphaelsc commented on locator/tablets.cc:
```diff
@@ -159,6 +159,17 @@ bool tablet_has_excluded_node(const locator::topology& topo, const tablet_info&
+    auto remainder = substract_sets(a.replicas, b.replicas);
+    if (!remainder.empty()) {
+        return {};
+    }
```
are you talking about a transient state where we have increased RF and one of the tables have more replicas than other? makes sense.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -977,13 +1136,22 @@ class load_balancer {
+            } else if (tmap.needs_merge() && plan.resize_plan().all_colocated.contains(table) && !bypass_merge_completion()) {
+                finalize_decision();
+                lblogger.info("Finalizing resize decision for table {} as all replicas are co-located",
+                              table, table_stats->split_ready_seq_number);
```
oops.

raphaelsc commented on service/tablet_allocator.cc:
```diff
@@ -880,7 +881,165 @@ class load_balancer {
+
+                auto r1 = get_replicas(t1);
+                auto r2 = get_replicas(t2);
+                if (r1 == r2) {
```
don't we need to restrict co-location migrations within a DC?

for example, let's say
DC1{ n1, n3 }
DC2{ n2, n4 }

tablet0 { n1, n4 }
tablet1 { n2, n3 }

we cannot migrate t1:n2 into t0:n1, since they're on different DCs.

if the problem is that nodes is not a good source of truth, I think I can instead rely on topology that can tell me if a particular host id lives in the target DC.

the decision should still take a look at skip list to avoid emitting a co-location migraton if either the source or destination is down.

raphaelsc replied:
I am not exactly sure what you meant by "This should be computed not in per-DC plan, but the global plan.". Maybe you still want the DC aware co-location (which is what you proposed in the original design), but want a second pass over the replica sets when determining that everything is co-located?

Isn't it enough if I fix get_replicas() to not discard any replica within the DC? then I can combine all_colocated from all resize plans (as happens in the current version).

tgrabiec replied:
> I am not exactly sure what you meant by "This should be computed not in per-DC plan, but the global plan.". Maybe you still want the DC aware co-location (which is what you proposed in the original design), but want a second pass over the replica sets when determining that everything is co-located?

Yes, I only meant that all_colocated should be moved to the global plan making. Co locating migration scheduling should be per DC. 

> 
> Isn't it enough if I fix get_replicas() to not discard any replica within the DC? then I can combine all_colocated from all resize plans (as happens in the current version).

It seems more complicated than necessary to maintain that state. It's also potentially unsafe because it relies on the assumption that we invoked plan maker for all DCs present in the replica list. I'm not sure it will always hold in the future.

raphaelsc commented:
Changes in v7:

- Rebased (resolved conflicts)
- Added cluster feature for tablet merge
- Fix co-location check in merge_tablet_info()
- Move co-location check to global plan making, co-location migration scheduling remains per DC.
- improve co-location logging by dumping global_tablet_id instead of plain tablet_id

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13573/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13573/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_merge
#### Failed Tests (40/1035):
- [test_tablet_merge_simple.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13573/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.1)
- [test_tablet_merge_simple.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13573/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.10)
- [test_tablet_merge_simple.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13573/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.2)
- [test_tablet_merge_simple.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13573/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.3)
- [test_tablet_merge_simple.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13573/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.4)
- [test_tablet_merge_simple.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13573/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.5)
- [test_tablet_merge_simple.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13573/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.6)
- [test_tablet_merge_simple.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13573/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.7)
- [test_tablet_merge_simple.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13573/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.8)
- [test_tablet_merge_simple.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13573/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_merge_simple_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_merge_simple.debug.9)
#### Build Details:
- Duration: 2 hr 10 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13599/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13599/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_merge
#### Failed Tests (1/1035):
- [test_tablet_split_and_merge_with_concurrent_topology_changes.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13599/testReport/junit/topology_experimental_raft/test_tablets_merge/test_tablet_split_and_merge_with_concurrent_topology_changes_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_split_and_merge_with_concurrent_topology_changes.dev.1)
#### Build Details:
- Duration: 1 hr 20 min
- Builder: spider3.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13605/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13605/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_merge
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/516/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5164/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13180/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2055/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13605/)
#### Build Details:
- Duration: 10 hr
- Builder: spider7.cloudius-systems.com

raphaelsc commented:
@tgrabiec please review and consider merging the merge :-)
